jgrapht-core/src/test/java/org/jgrapht/alg/TransitiveReductionTest.java

```diff
@@ -137,6 +138,77 @@ public void testReduceNull()
         TransitiveReduction.INSTANCE.reduce(null);
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void testCyclicGraph()
+    {
+        SimpleDirectedGraph<String, DefaultEdge> graph =
+          new SimpleDirectedGraph<>(DefaultEdge.class);
+        graph.addVertex("A");
+        graph.addVertex("B");
+        graph.addVertex("C");
+        graph.addVertex("D");
+        graph.addEdge("A", "B");
+        graph.addEdge("B", "C");
+        graph.addEdge("C", "D");
+        graph.addEdge("D", "A");
+        TransitiveReduction.INSTANCE.reduce(graph);
+    }
+
+    @Test
+    public void testAcyclicGraphNoCheck()
+    {
+        SimpleDirectedGraph<String, DefaultEdge> graph =
+          new SimpleDirectedGraph<>(DefaultEdge.class);
+        graph.addVertex("A");
+        graph.addVertex("B");
+        graph.addVertex("C");
+        graph.addEdge("A", "B");
+        graph.addEdge("B", "C");
+        graph.addEdge("A", "C");
+        assertFalse(new CycleDetector<>(graph).detectCycles());
+        TransitiveReduction.INSTANCE.reduce(graph, false);
+        int correctEdgeCount = 2;
+        assertEquals(correctEdgeCount, graph.edgeSet().size());
+    }
+
+    @Test
+    public void testCyclicGraphNoCheck()
+    {
+        SimpleDirectedGraph<String, DefaultEdge> graph =
+          new SimpleDirectedGraph<>(DefaultEdge.class);
+        graph.addVertex("A");
+        graph.addVertex("B");
+        graph.addVertex("C");
+        graph.addEdge("A", "B");
+        graph.addEdge("B", "C");
+        graph.addEdge("C", "A");
+        TransitiveReduction.INSTANCE.reduce(graph, false);
+        // CAVEAT: See how dangerous it is to call reduce(Graph, boolean) without the default cyclicity check?
```

Comment:
The warning is worthwhile, but a test verifying incorrect behavior shouldn't be included.  Maybe just move the warning to testAcyclicGraphNoCheck and invite the reader to investigate what happens if A->C is reversed.

Reply:
I disagree for two reasons:
  1. Tests for me do not just verify behaviour but are also a readable specification. When not forcing myself to use JUnit but in my own projects using the BDD tool Spock, test classes even inherit from a class called `Specification`.
  2. Verifying faulty behaviour in this case is helpful because if one day someone fixes the class under test or unifies the simple and directed transitive reduction classes into one (which I actually considered before deciding that I would leave that up to the maintainers), this test would fail and could be removed along with the comment.

Reply:
1. That's great, but it's meaningless to specify faulty behavior, unless you're working on some bug-for-bug-compatible emulator for a piece of legacy software.

2. Why wait when you can just remove it now?

Reply:
No, it is not meaningless because I am also **documenting** faulty behaviour that way. The test shows a use case in which the user can unintentionally mess up, using the parameter in the wrong situation.

Reply:
Have fun with that.

Reply:
This has stopped being fun quite a while ago. Anyway, thanks, I take your remark as an OK and will click "resolve" on this one.